src/connection.js

```diff
@@ -0,0 +1,126 @@
+import io from 'socket.io-client'
+import {
+  AMIO_WEBCHAT_SERVER_URL,
+  DEFAULT_LOCAL_STORAGE_SESSION_NAME,
+  SOCKET_CONNECTION_ACCEPTED,
+  SOCKET_CONNECTION_REJECTED,
+  SOCKET_IO_DISCONNECT,
+  SOCKET_IO_ERROR,
+  SOCKET_MESSAGE_SERVER,
+  SOCKET_NOTIFICATION_SERVER,
+  SOCKET_MESSAGE_ECHO,
+  ERROR_CODE_CHANNEL_ID_CHANGED
+} from './constants'
+
+class Connection {
+
+  constructor() {
+    this.storage = window.localStorage
+    if(!this.storage) {
+      // for tests
+      this.storage = {}
+      this.storage.getItem = () => {}
+      this.storage.setItem = () => {}
+    }
+    this.sessionId = null
+
+    this.messageReceivedHandler = () => {}
+    this.messageEchoHandler = () => {}
+    this.notificationReceivedHandler = () => {}
+    this.connectionStateChangedHandler = () => {}
+  }
+
+  connect(config) {
+    return new Promise((resolve, reject) => {
+      if(!config || !config.channelId) {
+        reject('Could not connect: config.channelId is invalid.')
+        return
+      }
+
+      // for dev purposes: set config._amioWebchatServerUrl to use a different server
+      const serverUrl = config._amioWebchatServerUrl || AMIO_WEBCHAT_SERVER_URL
+      const sessionName = config.localStorageSessionName || DEFAULT_LOCAL_STORAGE_SESSION_NAME
+
+      const opts = {
+        secure: true,
+        reconnection: true,
+        reconnectionDelay: 1000,
+        reconnectionDelayMax: 5000,
+        reconnectionAttempts: 99999,
+        query: {
+          v: 1,
+          channel_id: config.channelId
+        }
+      }
+
+      this.sessionId = this.storage.getItem(sessionName)
+      if(this.sessionId) {
+        opts.query.session_id = this.sessionId
+      }
+      this.socket = io(serverUrl, opts)
+
+      this.socket.on(SOCKET_CONNECTION_ACCEPTED, data => {
+        const sessionId = data.session_id
+
+        this.sessionId = sessionId
+        this.storage.setItem(sessionName, data.session_id)
+        this.connectionStateChangedHandler(true)
+        resolve()
+      })
+
+      this.socket.on(SOCKET_CONNECTION_REJECTED, data => {
+        if(data.error_code === ERROR_CODE_CHANNEL_ID_CHANGED) {
+          console.warn('Session invalidated by the server. New session will be created automatically.')
+          this.storage.removeItem(sessionName)
+          this.connect(config)
```

Comment:
This event happens only during the connection process OR it can happen while the chat-client is already connected?

Reply:
Only during connection (unless we have some serious bug somewhere)

Reply:
If I searched well, **socket.io** is really using [EventEmitter](https://nodejs.org/api/events.html#events_emitter_on_eventname_listener). The docs say:

> Adds the `listener` function to the end of the listeners array for the event named `eventName`. No checks are made to see if the listener has already been added. Multiple calls passing the same combination of `eventName` and `listener` will result in the `listener` being added, and called, multiple times.

That means, with current implementation we would have to [manually remove the **on** listeners](https://nodejs.org/api/events.html#events_emitter_removelistener_eventname_listener).

Reply:
If during connection we receive `ERROR_CODE_CHANNEL_ID_CHANGED` than what happens with the first `socket`. It seems like there will be 2 open connections, right?

Reply:
Added `socket.off()` before we re-connect to make it cleaner.